### ISSUE TITLE
SAMLTracer.js Error in Firefox Developer Edition (103.0b2)

### ISSUE THREAD

**Issue by kellenmurphy** [simplesamlphp/SAML-tracer#74]
SAML-tracer appears to have stopped functioning in the latest Firefox Beta (Developer Edition tracks with Beta)

```
Uncaught Error: Type error for parameter extraInfoSpec (Error processing 2: Invalid enumeration value "extraHeaders") for webRequest.onBeforeSendHeaders.
    init moz-extension://21213b32-20d7-404e-b9f3-25e9c5930f22/src/SAMLTrace.js:943
    <anonymous> moz-extension://21213b32-20d7-404e-b9f3-25e9c5930f22/src/ui.js:15
[SAMLTrace.js:943:42](moz-extension://21213b32-20d7-404e-b9f3-25e9c5930f22/src/SAMLTrace.js)
    makeError resource://gre/modules/Schemas.jsm:539
    throwError resource://gre/modules/Schemas.jsm:2659
    fixedArgs resource://gre/modules/Schemas.jsm:2727
    map self-hosted:180
    checkParameters resource://gre/modules/Schemas.jsm:2720
    addStub resource://gre/modules/Schemas.jsm:2956
    init moz-extension://21213b32-20d7-404e-b9f3-25e9c5930f22/src/SAMLTrace.js:943
    <anonymous> moz-extension://21213b32-20d7-404e-b9f3-25e9c5930f22/src/ui.js:15
```

I've tried tried disabling all other extensions, etc. This occurs only with FF 103 insofar as I can tell... which will be the "main" version of Firefox come July 22. I don't know precisely that this is an issue with SAML-tracer vs. FF, though a cursory search on Bugzilla doesn't reflect anything related to changes/bugs with `webRequest.onBeforeSendHeaders`.

Screenshot:
![2022-06-30 10_31_28](https://user-images.githubusercontent.com/30330479/176703942-206c3066-b592-435f-ae05-e4c1eb4fd507.png)

Could any of the maintainers provide some insight? I typically use developer edition just so that I have a completely separate browser for debugging SAML (hate clearing cache on my main browser). I can work around this tentatively, but I'd hate to see FF103 drop in July and break SAML-Tracer.

**Comment by thijskh** [comment]
I think this is where it goes awry:
https://github.com/simplesamlphp/SAML-tracer/blob/a500371f614c07f4b86871e81f7bcc2e46a82ee9/src/SAMLTrace.js#L907-L913

Apparently your browser does not satisfy the test in the method "isFirefox()". So a value is sent that Firefox does not understand. Question is why your Firefox is not Firefox.

**Comment by thijskh** [comment]
[This Bugzilla item seems](https://bugzilla.mozilla.org/show_bug.cgi?id=1772905) to describe more about the InstallTrigger deprecation and explicitly mentions the use of this construct to test for Firefoxness. And if understand it well, also that in "late beta" they will make using InstallTrigger for this purpose work again?

**Comment by tvdijen** [comment]
We might consider doing an actual user agent check instead of using this strange `InstallTrigger` global?

**Comment by khlr** [comment]
This is a really interesting issue! 🙂👍

I think you have to ask yourself why `isFirefox()` is needed at all. It was intruduced with pull request #61 because Chrome 72+ requires `extraHeaders` to access some certain HTTP headers. Therefore, "is Firefox" is actually misleading. It is rather the question whether it "is Chrome". 

We could easily utilize the user agent string to determine the browser type. By doing something like `navigator.userAgent.indexOf("Chrome") !== -1`. But [user agent detection should be avoided](https://developer.mozilla.org/en-US/docs/Web/HTTP/Browser_detection_using_the_user_agent#avoiding_user_agent_detection) if possible.

Instead we could use feature detection and check if the "feature" "extraHeaders" is available. Probably by doing something like this:

```js
var browser = browser || chrome;
const useExtraHeaders = () => {
  try  {
    const dummyListener = req => console.log(req);
    browser.webRequest.onBeforeSendHeaders.addListener(dummyListener, {urls: ["<all_urls>"]}, ["extraHeaders"]);
    browser.webRequest.onBeforeSendHeaders.removeListener(dummyListener);
    // seems to be Chrome which supports an requires "extraHeaders"
    return true;
  } catch (err) {
    // seems to be Firefox since "extraHeaders" is not supported
    return false;
  }
}

console.log(useExtraHeaders());
```

But I'm afraid this is a pretty expensive piece of code to check for the feature... Any ideas?

**Comment by tvdijen** [comment]
Is it actually still a problem to pass the extra parameter in Firefox? Because it is in the [Mozilla MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onBeforeSendHeaders) and nothing on that page indicates it wouldn't work on Firefox.. Is it perhaps something from the past?

If we still need it, it should perhaps be possible to test the amount of parameters? Like `browser.webRequest.onBeforeSendHeaders.length`? That would work for methods, but I'm not sure it will also work for events..

**Comment by khlr** [comment]
The problem isn't `addListener`'s third parameter (`extraInfoSpec`). It's rather the values that `extraInfoSpec` can hold.
While the Chrome-webRequest-API _requires_ the value `extraHeaders` (see [here](https://developer.chrome.com/docs/extensions/reference/webRequest/)), the Firefox-webRequest-API can't handle that value (the docs don't mention it at all) and FF throws the exception that Kellen reported: _Invalid enumeration value "extraHeaders"_

**Comment by tvdijen** [comment]
Ah, my bad.. Forget what I said :)

**Comment by thijskh** [comment]
Ideally we'd change the detection method. But I would only change it if it's actually a significant improvement. The code suggested by @khlr seems indeed quite a lot of code/effort for the simple case.

They are adding the object back in late beta (even if it's only null) precisely because the way we test for it was "the way to do it" and many extensions did it. So I'm not afraid it will break quickly in the future. It would seem a premature optimization to replace this simple check with complex code if the simple check is not likely to go away for the forseeable future.

So long story short, I'd like to replace it only if there's a similarly straightforward check available.

**Comment by tvdijen** [comment]
How about the `window.chrome` var?

**Comment by kellenmurphy** [comment]
I think there may be a predefined constant can be leveraged here (see: https://stackoverflow.com/questions/69287721/how-to-manipulate-webrequest-cookie-in-a-cross-browser-extension).

Specifically, the answer:

>The API exposes all predefined constants in chrome.webRequest.OnXXXXXXXXX objects for each event so only in new Chrome such objects will have EXTRA_HEADERS key with extraHeaders value whereas in Firefox and old Chrome it'll be undefined, which can be filtered out via filter()

```
chrome.webRequest.onBeforeSendHeaders.addListener(
  listenerFunc,
  { urls: ['*://*/*'] },
  ['blocking', 'requestHeaders', 
   chrome.webRequest.OnBeforeSendHeadersOptions.EXTRA_HEADERS].filter(Boolean)
);
```

I'll try to test this... would be happy to create a PR with that change if it works. 

It seems to me that a simple check that can be applied with initiating the listener seems like the simplest solution long term, and potentially the new "right way" to do things once `InstallTrigger` is removed.

**Comment by khlr** [comment]
Thank you for this contribution, Kellen! 🙂
It's good to have this change in the codebase!

However, I think we can do without a release for now, since `InstallTrigger` is being made available again.
